enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,347 @@
+
+The mixed-cpu-node-plugin itself is essentially a pod that runs as a DaemonSet on the worker nodes.
+cluster-node-tuning-operator (NTO) will be modified to support the deployment of the DaemonSet and additional
+resources (ServiceAccount, RBAC resources, SecurityContextConstraint, etc.)
+NTO will also be responsible to watch, monitor and report the node-plugin's DaemonSet status.
+Additionally, since the plugin is part of OCP infrastructure pods, NTO annotates the pods with:
+```yaml
+target.workload.openshift.io/management: {
+  "effect":"PreferredDuringScheduling"
+}
+```
+See https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md for details.
+
+Since both the components are related,
+mixed-cpu-node-plugin [source code](https://github.com/openshift-kni/mixed-cpu-node-plugin) would be vendor under NTO.
+By having different main.go + container entrypoint, NTO image can be also used as the
+mixed-cpu-node-plugin image. 
+This way saves the overhead of having a separate container image with all the maintenance cost that comes with it.
+
+In order to activate NRI in container-runtime, NTO creates MachineConfig to enable NRI in 
+the container-runtime configuration.
+
+The plugin is completely optional and is off by default.
+
+No changes are needed in the container-runtime, because NRI hooks already supported by the runtime
+and mixed-cpu-node-plugin implementing the NRI interface.
+This minimizes risks and helps avoid regression on
+a container-runtime component (CRI-O in OpenShift case).
+
+### API
+
+Extend PerformanceProfile with new `workloadHints` named `mixedCpus` in order to enable
+the feature.
+
+```yaml
+workloadHints:
+  mixedCpus: true
+```
+
+Specify the shared cpus via annotation:
+
+```yaml
+annotations:
+    mixed-cpu.openshift.io/shared-cpus: "2-3"
+```
+
+The CPU set value should follow the 
+[cpuset](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpuset) conventions.
+It defines a set of CPUs that can be allocated to guaranteed pods/containers that still require
+non-exclusive ones.
+
+The CPU set must not overlap with `spec.cpu.reserved` or `spec.cpu.isolated` and NTO
+should return an error in case it does.
+
+The reason for using an annotation and not new API field 
+is because there is a coming feature named shared-partition
+that uses similar convention i.e., shared cpus, and would be 
+enabled via NTO as well.
+
+Once shared-partition lands, 
+the annotation should be [deprecated](mixed-cpu-node-plugin.md#removing-a-deprecated-feature)
+and node-plugin will use the shared cpus specified 
+at `spec.cpu.shared`.
+
+Both workloadHints and the annotation have to be specified in order to activate the feature.
+If only one of them would be specified, NTO should report a warning (in logs or status). 
```

Comment:
If this is a Must I wonder if we can enforce via validation that if you explicitly specify the annotation then you must have the workloadhints specified.
Lets ask it this way -  if the annotation alone is specified will anything happen?

Reply:
Nothing should happen.

Reply:
What information is provided in the hints vs the annotation? From an API perspective, if these are joined and need to be set in tandem, then they should be API fields close together with validation to ensure they must both be set.

Having them disjoint is likely to lead to confusion for end users